### ISSUE TITLE
np.load, np.save missing

### ISSUE THREAD

**Issue by Fokatu** [Quansight-Labs/numpy.net#48]
Is it possible to add these two functions to load and save npy file?
Thanks!

**Comment by KevinBaselinesw** [comment]
I think I looked at np.load and np.save before.  It is a very big task to implement these and I didn't have time to do it.  Nobody has asked for them in 5 years so they are not the most commonly used functions.

In python/numpy there is a very significant performance improvement by saving/loading the arrays with those functions because they are written in C and not python.  My view is that an application user of numpydotnet should probably write their own specific load/save functions.   Since it would be in .NET either way, there is no performance to be gained by replicating these functions in the library.

Can you implement your own load and save functions?  
Does the format need to be 100% compatible with python/numpy?
Are you exchanging files with python/numpy?

FYI, we do support np.tofile and np.fromfile

If you really need to be 100% compatible with the python file format, let me know.
I can also offer tips on accessing any arrays you want to save to file.

Also, we recently added support for serializing ndarrays.   See ndarray.ToSerializable().   It will return a class object that can be serialized to a string and written to a disk file.

**Comment by rainyl** [comment]
I think it is **not necessary** to add `.npy` support for `numpy.net`, even as a python numpy user, I hardly use `np.save` or `np.load`, I prefer text file like `.csv`, if higher performance needed, I think `feather` is a better choice.

**Comment by rainyl** [comment]
@KevinBaselinesw I noticed that you are working on `np.load` for binary files, an advice, text files and binary files can use different api, for example, numpy use `np.loadtxt` for text file and `np.load` for binary file, which will be more clear and direct. 

PS: the current `np.fromfile` in this project is also not complete, e.g., `skip_rows`, `comments` are useful when the input file has extra info, but they are not included. 

Finally, I REALLY appreciate your works, I explored many multi-dimensional array libraries for C#, such as NumSharp, Numpy.NET, Tensor.NET from SciSharp, thanks for their efforts, but they are not convenient and complete as this work, TorchSharp is great but I don't want to install such a huge library just for multi-dimension array calculation.

Thanks for your work again!!! 😄

**Comment by rainyl** [comment]
Yes, you are right, for saving files, `.ToSerializable()` is a better choice. Considering the complexity, I think it's not necessary to support `np.load`, at least with lower priority. Anyway, it's only a immature suggestion from a new user, thanks for your reply :)

**Comment by GeorgeS2019** [comment]
@KevinBaselinesw 

Could you evaluate if these codes are useful "**Starting Ideas**" to address the missing np.load and np.save functions

https://github.com/SciSharp/NumSharp/blob/master/src/NumSharp.Core/APIs/np.load.cs

Less relevant, but FYI
https://github.com/SciSharp/Numpy.NET/blob/main/src/Numpy/np.io.gen.cs

**Comment by thalesfm** [comment]
I've been working on implementing `np.load` for a project of mine since I needed access to a dataset that uses the `.npz` format. It's not a 1-to-1 port since I don't handle structured data types, but I've gotten most of the primitive types working so far. Would you be open to include this feature if I submit a pull request?

**Comment by GeorgeS2019** [comment]
submit the PR, it will be useful for the community, especially those from the **Scisharp**

**Comment by KevinBaselinesw** [comment]
sorry I was out of town for a few days.

Feel free to create a PR (with some unit tests please).  I will review it and potentially accept it.

**Comment by williamlzw** [comment]
I am porting a python version of onnx inference project to c#.
This repository is great, but it lacks np.load function to load the necessary npy files.
The npy file link is here: https://huggingface.co/lovemefan/SenseVoice-onnx/blob/main/embedding.npy

https://github.com/lovemefan/SenseVoice-python/blob/main/sensevoice/onnx/sense_voice_ort_session.py
line132:  self.embedding = np.load(embedding_model_file)

**Comment by williamlzw** [comment]
![Image](https://github.com/user-attachments/assets/ac302d4c-c10a-4275-95ee-45047cf6cb18)

**Comment by KevinBaselinesw** [comment]
@williamlzw Unfortunately I never got the np.load function to be fully implemented.  Ultimately, we decided a better .NET way was to serialize the ndarray objects and save them using JSON or XML.  (see the notes earlier in this thread).  That does impact any user who is trying to be compatible with an existing python implementation that is loading npy files though.  Below is the code that I started writing.  Hopefully you can use it as a starting point and implement the feature to your needs.

```
  /// <summary>
    /// implements functions similar to npyio.py
    /// </summary>
    public static partial class np
    {
        static byte[] _ZIP_PREFIX = new byte[] { 0x50, 0x4B, 0x03, 0x04 }; // b'PK\x03\x04';
        static byte[] MAGIC_PREFIX = new byte[] { 0x93, 0x4e, 0x55, 0x4d, 0x50, 0x59 }; // b'\x93NUMPY';
        static int MAGIC_LEN = MAGIC_PREFIX.Length + 2;

        public static ndarray load(string PathName)
        {
            throw new Exception("This function is under consideration for development");

            if (string.IsNullOrEmpty(PathName))
            {
                throw new Exception("Pathname null or empty");
            }

            if (System.IO.File.Exists(PathName) == false)
            {
                throw new Exception("Specified file does not exist!");
            }

            var fp = System.IO.File.Open(PathName, System.IO.FileMode.Open);


            int N = MAGIC_PREFIX.Length;

            byte[] magic = new byte[N + 200];

            var magic_read = fp.Read(magic, 0, N);

            fp.Seek(-Math.Min(N, magic_read), System.IO.SeekOrigin.Current);

            if (IsPrefixMatch(magic, _ZIP_PREFIX))
            {
                throw new Exception("Zipped files are not supported");
            }
            if (IsPrefixMatch(magic, MAGIC_PREFIX))
            {
                return read_array(fp);
            }
            return null;
        }

        private static ndarray read_array(FileStream fp)
        {
            var version = read_magic(fp);
            _check_version(version);

            var array_info = _read_array_header(fp, version);


            throw new NotImplementedException();
        }


        private static object _read_array_header(FileStream fp, (int major, int minor) version)
        {
            int struct_calcsize = -1;
            if (version.major == 1 && version.minor == 0)
            {
                struct_calcsize = 2;
            }
            else if (version.major == 2 && version.minor == 0)
            {
                struct_calcsize = 4;
            }
            else
            {
                throw new ValueError(string.Format("Invalid version ({0},{1})" + version.major, version.minor));
            }


            var hlength_str = _read_bytes(fp, struct_calcsize, "array header length");

            var header = _read_bytes(fp, hlength_str[0], "array_header");

            throw new NotImplementedException();
        }

        private static void _check_version((int major, int minor) version)
        {
            if (version.major == 1 && version.minor == 0)
                return;
            if (version.major == 2 && version.minor == 0)
                return;

            throw new Exception(string.Format("we only support version (1,0) and (2,0), not ({0},{1})", version.major, version.minor));
        }

        private static (int major, int minor) read_magic(FileStream fp)
        {
            byte[] magic_str = _read_bytes(fp, MAGIC_LEN, "magic string");
            if (IsPrefixMatch(magic_str, MAGIC_PREFIX) == false)
            {
                throw new Exception("File does not have expected magic string");
            }

            int major = magic_str[MAGIC_LEN - 2];
            int minor = magic_str[MAGIC_LEN - 1];

            return (major, minor);
        }

        private static byte[] _read_bytes(FileStream fp, int size, string error_template)
        {

            byte[] data = new byte[size];
            int data_offset = 0;

            while (true)
            {
                try
                {
                    int r = fp.Read(data, data_offset, size - data_offset);
                    data_offset += r;
                    if (data_offset == size)
                        break;
                }
                catch (Exception ex)
                {
                    string msg = string.Format("EOF: reading {0}, expected {1} bytes got {2}", error_template, size, data_offset);
                    throw new Exception(msg);
                }
            }

            return data;
        }

        private static bool IsPrefixMatch(byte[] magic, byte[] Prefix)
        {
            for (int i = 0; i < Prefix.Length; i++)
            {
                if (magic[i] != Prefix[i])
                    return false;
            }

            return true;
        }

    }
```

**Comment by KevinBaselinesw** [comment]
google "numpy np.load source code" to get a look at the original source code.

**Comment by KevinBaselinesw** [comment]
From earlier, this competing library might offer some tips:

https://github.com/SciSharp/NumSharp/blob/master/src/NumSharp.Core/APIs/np.load.cs

Less relevant, but FYI
https://github.com/SciSharp/Numpy.NET/blob/main/src/Numpy/np.io.gen.cs

Also, it appears that https://github.com/thalesfm was working on implementing np.load for his/her own needs.  Perhaps they would be willing to share.

**Comment by thalesfm** [comment]
@KevinBaselinesw Yes, I have an implementation of `np.load` that works for most primitive data types (apologies for not getting around to submitting that PR I mentioned earlier!). My implementation reads the file into a native C# array and outputs the shape, byte order, etc. separately, so I suspect it integrating it into NumpyDotNet would just be a matter of calling the appropriate ndarray constructor with the data. I tried using `Buffer.BlockCopy` whenever possible, so the whole thing ended up being reasonably fast too.

I could submit a PR with the code not problem, though I would appreciate some help integrating it into the project.

**Comment by KevinBaselinesw** [comment]
@thalesfm I will definitely help you integrate this into numpydotnet.  I want to make sure you get credited for it to.

But first, could we get your code to @williamlzw to see if it works for their problem?  I always try to get people unstuck as soon as possible.  Can you post your code here?  Or does @williamlzw have an email address?   @thalesfm thank you for your help!

**Comment by thalesfm** [comment]
@KevinBaselinesw The whole thing is available at https://github.com/thalesfm/unity-nerf/tree/master/Packages/xyz.example.nerf/Editor/IO. It's part of a Unity project I'm currently working on, but the code should work fine on it's own (except for the conversion from half-precision and single-precision floats, since I'm using a Unity function for that).

That repo doesn't have a license attached to it yet, but if @williamlzw wants to use the code just let me know and I'll add one.

**Comment by williamlzw** [comment]
I am temporarily using this method to read data.

![Image](https://github.com/user-attachments/assets/31615370-69ef-4273-882c-902f55ce1af1)

**Comment by KevinBaselinesw** [comment]
https://www.nuget.org/packages/NumpyDotNet/0.9.87 has an implementation of np.load and np.save.   @thalesfm thank you very much for your sample.  It saved many hours of work figuring out and coding it.  @williamlzw

**Comment by thalesfm** [comment]
@KevinBaselinesw No problem.

I noticed your implementation throws an exception when the array is in Fortran order, but I suspect you should be able to handle that by simply transposing the ndarray before returning it and everything should to work fine. (The only reason I didn't do that is because it would be annoying to transpose a native C# array and I was feeling lazy :))

**Comment by KevinBaselinesw** [comment]
I have never paid much attention to fortran ordering.  I don't know if anyone is even using it in NumpyDotNet.  But all I really needed to do was specify the fortran order flag when I recreate the ndarray object.  I will wait a few days before doing a new release in case anyone reports any issues.

**Comment by williamlzw** [comment]
Version 0.9.87 now reads npy files correctly. Can you provide a nuget package for dotnet8.0 because I write deep learning code such as TorchSharp that depends on at least dotnet6.0 or dotnet8.0.

**Comment by KevinBaselinesw** [comment]
@williamlzw The numpydotnet library is built for .net standard 2.0.   I don't claim to be an expert and the number of versions of .NET has become very large and is difficult to keep track of.  But my understanding is that .netstandard 2.0 was compatible with almost everything including .net 8.0.  Are you sure that the library as built can't be used in your project?

If you need a specific version like that, I wonder if it might be best to grab a copy and rebuild it to your needs.  It is open source.

I am not opposed to building it in multiple versions, but I am not sure it is necessary and quite frankly I am not sure how to go about publishing multiple versions.

**Comment by williamlzw** [comment]
np.array() cannot parse textual data such as {"-8.0123","5.331"} to float64 arrays.

https://huggingface.co/lovemefan/SenseVoice-onnx/blob/main/fsmn-am.mvn
```
cmvn_file = "D:\\model\\sensevoice-onnx\\fsmn-am.mvn"
with open(cmvn_file, "r", encoding="utf-8") as f:
            lines = f.readlines()

means_list = []
vars_list = []
for i in range(len(lines)):
    line_item = lines[i].split()
    if line_item[0] == "<AddShift>":
        line_item = lines[i + 1].split()
        if line_item[0] == "<LearnRateCoef>":
            add_shift_line = line_item[3 : (len(line_item) - 1)]
            means_list = list(add_shift_line)
            continue
    elif line_item[0] == "<Rescale>":
        line_item = lines[i + 1].split()
        if line_item[0] == "<LearnRateCoef>":
            rescale_line = line_item[3 : (len(line_item) - 1)]
            vars_list = list(rescale_line)
            continue

means = np.array(means_list).astype(np.float64)
print(means)
```

**Comment by KevinBaselinesw** [comment]
Yes, of course we can't automatically convert a string to a double.  Does that work in python?

**Comment by williamlzw** [comment]
The C# version does not support array[[[0]]] indexing.
![Image](https://github.com/user-attachments/assets/a18c3502-68ca-450f-bc2e-a375fb6beb2f)

**Comment by KevinBaselinesw** [comment]
```
    [TestMethod]
        public void williamlzw_2()
        {
            ndarray s1 = np.array(new string[] { "1.1", "2.2", "3.3" });

            ndarray s2 = (ndarray)s1[(new long[] { 1 })];
            // s2 = ndarray("2.2");

            string s3 = (string)s1[ 1 ];
            // s3 = string == "2.2"

        }
```


I think this is what you are trying to do.  This kind of indexing requires an array of longs.

**Comment by williamlzw** [comment]
var languageQuery = ((ndarray)_embedding[language]).reshape(1, 1, -1);
I solved it using this method.
